File: union-find-medium/130.被围绕的区域.cpp
/*
 * @lc app=leetcode.cn id=130 lang=cpp
 *
 * [130] 被围绕的区域
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class UnionFind {
private:
    int count;
    vector<int> parent, rank;

public:
    UnionFind(int n) {
        count = n;
        parent.resize(n);
        rank.resize(n);
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 1;
        }
    };

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void unionn(int x, int y) {
        int rootx = find(x), rooty = find(y);
        if (rootx != rooty) {
            if (rank[rootx] > rank[rooty]) {
                swap(rootx, rooty);
            }
            parent[rootx] = rooty;
            if (rank[rootx] == rank[rooty]) {
                rank[rooty]++;
            }
            count--;
        }
    }
};


class Solution {
public:
    int row, col;
    void solve(vector<vector<char>> &board) {
        row = board.size(), col = board.size() == 0 ? 0 : board[0].size();
        if (row == 0 || col == 0) {
            return;
        }

        UnionFind uf(row * col + 1);
        int dummy = row * col;

        auto node = [&](int i, int j) {
            return i * col + j;
        };

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (board[i][j] != 'O') continue;
                if ((i == 0 || i == row - 1 || j == 0 || j == col - 1)) {
                    uf.unionn(node(i, j), dummy);
                } else {
                    if (i - 1 > -1 && board[i - 1][j] == 'O')
                        uf.unionn(node(i, j), node(i - 1, j));
                    if (i + 1 < row && board[i + 1][j] == 'O')
                        uf.unionn(node(i, j), node(i + 1, j));
                    if (j - 1 > -1 && board[i][j - 1] == 'O')
                        uf.unionn(node(i, j), node(i, j - 1));
                    if (j + 1 < col && board[i][j + 1] == 'O')
                        uf.unionn(node(i, j), node(i, j + 1));
                }
            }
        }

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (uf.find(node(i, j)) == uf.find(dummy)) {
                    board[i][j] = 'O';
                } else {
                    board[i][j] = 'X';
                }
            }
        }
    }
};
// @lc code=end
int main(int argc, char const *argv[]) {
    vector<vector<char>> v = {{'O', 'O', 'O'}, {'O', 'O', 'O'}, {'O', 'O', 'O'}};
    Solution().solve(v);
    return 0;
}

File: union-find-medium/130.被围绕的区域.py
#
# @lc app=leetcode.cn id=130 lang=python3
#
# [130] 被围绕的区域
#
from typing import List


# @lc code=start
class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n
        self.count = n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int):
        rootx, rooty = self.find(x), self.find(y)
        if rootx != rooty:
            if self.rank[rootx] > self.rank[rooty]:
                rootx, rooty = rooty, rootx
            self.parent[rootx] = rooty
            if self.rank[rootx] == self.rank[rooty]:
                self.rank[y] += 1
            self.count -= 1


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        row, col = len(board), 0 if len(board) == 0 else len(board[0])
        if row == 0 or col == 0:
            return

        def node(i: int, j: int) -> int:
            return i * col + j

        uf = UnionFind(row * col + 1)
        dummy = row * col
        for i in range(0, row):
            for j in range(0, col):
                if board[i][j] != 'O':
                    continue
                if i == 0 or i == row - 1 or j == 0 or j == col - 1:
                    uf.union(node(i, j), dummy)
                else:
                    for r, c in [(-1, 0), (0, 1), (1, 0), (0, -1)]:
                        x, y = r + i, c + j
                        if -1 < x < row and -1 < y < col and board[x][y] == 'O':
                            uf.union(node(i, j), node(x, y))

        for i in range(0, row):
            for j in range(0, col):
                if board[i][j] == 'O' and uf.find(node(i, j)) != uf.find(dummy):
                    board[i][j] = 'X'
# @lc code=end
